File: src/pythonnative/sdk/templates/contracts.swift
import Foundation
import CoreFoundation

/// Executable generated contracts. No schema JSON is interpreted on the mount path.
public enum PNContracts {
    public static let fingerprint = "{{fingerprint}}"
    private struct Field {
        let matches: (Any) -> Bool
        let allowed: Bool
        let layout: Bool
        let recreate: Bool
        let required: Bool
        let defaultValue: Any
        init(_ matches: @escaping (Any) -> Bool, _ allowed: Bool, _ layout: Bool, _ recreate: Bool, _ required: Bool, _ defaultValue: Any) {
            self.matches = matches; self.allowed = allowed; self.layout = layout
            self.recreate = recreate; self.required = required; self.defaultValue = defaultValue
        }
    }
    public struct ChangeMask: OptionSet {
        public let rawValue: Int
        public init(rawValue: Int) { self.rawValue = rawValue }
        public static let layout = ChangeMask(rawValue: 1)
        public static let recreate = ChangeMask(rawValue: 2)
    }
{{predicates}}
{{fields}}
    private static let components: [String: [String: Field]] = [
{{components}}
    ]
    private static let commands: [String: (Any) -> Bool] = [
{{commands}}
    ]
    private static let modules: [String: (Any) -> Bool] = [
{{modules}}
    ]
    private static func isBoolean(_ value: Any) -> Bool {
        (value as? NSNumber).map { CFGetTypeID($0) == CFBooleanGetTypeID() } ?? false
    }
    private static func isNumber(_ value: Any) -> Bool {
        (value as? NSNumber).map { CFGetTypeID($0) != CFBooleanGetTypeID() && $0.doubleValue.isFinite } ?? false
    }
    private static func isInteger(_ value: Any) -> Bool {
        guard isNumber(value), let number = value as? NSNumber else { return false }
        return abs(number.doubleValue) <= 9_007_199_254_740_991 && number.doubleValue.rounded() == number.doubleValue
    }
    public static func validate(_ name: String, _ props: [String: Any], partial: Bool = false) -> Bool {
        guard let fields = components[name] else { return false }
        if !partial && fields.contains(where: { $0.value.required && props[$0.key] == nil }) { return false }
        return props.allSatisfy { key, value in fields[key].map { $0.allowed && $0.matches(value) } ?? false }
    }
    public static func changes(_ name: String, _ keys: Set<String>) -> ChangeMask {
        guard let fields = components[name] else { return [.layout, .recreate] }
        var mask: ChangeMask = []
        for key in keys {
            if fields[key]?.layout ?? true { mask.insert(.layout) }
            if fields[key]?.recreate ?? false { mask.insert(.recreate) }
        }
        return mask
    }
    public static func invalidatesLayout(_ name: String, _ changed: [String: Any]) -> Bool {
        changes(name, Set(changed.keys)).contains(.layout)
    }
    public static func validateRemoval(_ name: String, _ changed: [String: Any], _ removed: [String]) -> Bool {
        guard let fields = components[name] else { return false }
        return Set(removed).count == removed.count && removed.allSatisfy { key in
            fields[key].map { !$0.required && $0.allowed && changed[key] == nil } ?? false
        }
    }
    public static func requiresRecreation(_ name: String, _ changed: [String: Any], removed: [String] = []) -> Bool {
        changes(name, Set(changed.keys).union(removed)).contains(.recreate)
    }
    public static func normalize(_ name: String, _ changed: [String: Any], removed: [String] = []) -> [String: Any] {
        var result = changed
        for key in removed { result[key] = components[name]?[key]?.defaultValue ?? NSNull() }
        return result
    }
    public static func validateCommand(_ name: String, _ method: String, _ args: [String: Any]) -> Bool {
        commands[name + "." + method]?(args) ?? false
    }
    public static func validateModule(_ name: String, _ method: String, _ args: [String: Any]) -> Bool {
        if ["Host", "Layout", "Runtime"].contains(name) { return true }
        return modules[name + "." + method]?(args) ?? false
    }
}

File: src/pythonnative/sdk/templates/contracts.kt
package com.pythonnative.generated

import org.json.JSONArray
import org.json.JSONObject

/** Executable generated contracts; the mount path doesn't interpret schema JSON. */
object PNContracts {
    const val fingerprint = "{{fingerprint}}"
    private data class Field(val matches: (Any) -> Boolean, val allowed: Boolean, val layout: Boolean,
                             val recreate: Boolean, val required: Boolean, val defaultValue: Any)
    const val LAYOUT = 1
    const val RECREATE = 2
{{predicates}}
{{fields}}
    private val components: Map<String, Map<String, Field>> = mapOf(
{{components}}
    )
    private val commands: Map<String, (Any) -> Boolean> = mapOf(
{{commands}}
    )
    private val modules: Map<String, (Any) -> Boolean> = mapOf(
{{modules}}
    )
    private fun isBoolean(value: Any): Boolean = value is Boolean
    private fun isNumber(value: Any): Boolean = value is Number && value.toDouble().isFinite()
    private fun isInteger(value: Any): Boolean = value is Number && value.toDouble().isFinite() &&
        kotlin.math.abs(value.toDouble()) <= 9007199254740991.0 && value.toDouble() == value.toLong().toDouble()
    fun validate(name: String, props: JSONObject, partial: Boolean = false): Boolean {
        val fields = components[name] ?: return false
        if (!partial && fields.any { (key, field) -> field.required && !props.has(key) }) return false
        return props.keys().asSequence().all { key -> fields[key]?.let { it.allowed && it.matches(props.get(key)) } ?: false }
    }
    fun changes(name: String, keys: Iterable<String>): Int {
        val fields = components[name] ?: return LAYOUT or RECREATE
        var mask = 0
        for (key in keys) {
            if (fields[key]?.layout != false) mask = mask or LAYOUT
            if (fields[key]?.recreate == true) mask = mask or RECREATE
        }
        return mask
    }
    fun invalidatesLayout(name: String, changed: JSONObject): Boolean = changes(name, changed.keys().asSequence().asIterable()) and LAYOUT != 0
    fun validateRemoval(name: String, changed: JSONObject, removed: List<String>): Boolean {
        val fields = components[name] ?: return false
        return removed.toSet().size == removed.size && removed.all { key -> fields[key]?.let { !it.required && it.allowed && !changed.has(key) } ?: false }
    }
    fun requiresRecreation(name: String, changed: JSONObject, removed: List<String> = emptyList()): Boolean =
        changes(name, changed.keys().asSequence().asIterable() + removed) and RECREATE != 0
    fun normalize(name: String, changed: JSONObject, removed: List<String> = emptyList()): JSONObject {
        val result = JSONObject()
        for (key in changed.keys()) result.put(key, changed.get(key))
        for (key in removed) result.put(key, components[name]?.get(key)?.defaultValue ?: JSONObject.NULL)
        return result
    }
    fun validateCommand(name: String, method: String, args: JSONObject): Boolean = commands["$name.$method"]?.invoke(args) ?: false
    fun validateModule(name: String, method: String, args: JSONObject): Boolean {
        if (name in setOf("Host", "Layout", "Runtime")) return true
        return modules["$name.$method"]?.invoke(args) ?: false
    }
}
